log4j2.disableJmx and/or log4j2.disable.jmx seem not to work

## Description

System property switches "log4j2.disableJmx" and/or "log4j2.disable.jmx" seem not to work!

## Configuration

**Version:** 2.23.1 with slf4j 1.7.36

**Operating system:** Linux Mint 21.2 Cinnamon

**JDK:** Azu Zulu Build of OpenJDK - Zulu17.52+17-CA 17.0.12+7-LTS, Java version 17.0.12

## Logs

```
2024-07-27T21:19:27.967946158Z main ERROR Could not reconfigure JMX java.lang.NoClassDefFoundError: javax/management/InstanceNotFoundException
	at org.apache.logging.log4j.core.LoggerContext.setConfiguration(LoggerContext.java:642)
	at org.apache.logging.log4j.core.LoggerContext.reconfigure(LoggerContext.java:713)
	at org.apache.logging.log4j.core.LoggerContext.reconfigure(LoggerContext.java:735)
	at org.apache.logging.log4j.core.LoggerContext.start(LoggerContext.java:260)
	at org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(Log4jContextFactory.java:154)
	at org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(Log4jContextFactory.java:46)
	at org.apache.logging.log4j.LogManager.getContext(LogManager.java:197)
	at org.apache.logging.log4j.spi.AbstractLoggerAdapter.getContext(AbstractLoggerAdapter.java:136)
	at org.apache.logging.slf4j.Log4jLoggerFactory.getContext(Log4jLoggerFactory.java:58)
	at org.apache.logging.log4j.spi.AbstractLoggerAdapter.getLogger(AbstractLoggerAdapter.java:46)
	at org.apache.logging.slf4j.Log4jLoggerFactory.getLogger(Log4jLoggerFactory.java:32)
	at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:363)
	at ch.autumo.ifacex.studio.IfaceXStudio.<clinit>(IfaceXStudio.java:77)
	at ch.autumo.ifacex.studio.Main.main(Main.java:26)
Caused by: java.lang.ClassNotFoundException: javax.management.InstanceNotFoundException
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	... 14 more

2024-07-27T21:19:28.390794174Z main ERROR Could not reconfigure JMX java.lang.NoClassDefFoundError: javax/management/InstanceNotFoundException
	at org.apache.logging.log4j.core.LoggerContext.setConfiguration(LoggerContext.java:642)
	at org.apache.logging.log4j.core.LoggerContext.reconfigure(LoggerContext.java:739)
	at org.apache.logging.log4j.core.config.Configurator.reconfigure(Configurator.java:263)
	at ch.autumo.commons.logging.Log4j2Logging.initializeInMemory(Log4j2Logging.java:184)
	at ch.autumo.commons.logging.Log4j2Logging.initializeInMemoryFile(Log4j2Logging.java:87)
	at ch.autumo.ifacex.studio.IfaceXStudio.main(IfaceXStudio.java:797)
	at ch.autumo.ifacex.studio.Main.main(Main.java:26)
Caused by: java.lang.ClassNotFoundException: javax.management.InstanceNotFoundException
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	... 7 more

```

Java-Code:
```
    public static void main(String[] args) {
        System.setProperty("log4j2.disable.jmx", "true");
        System.setProperty("log4j2.disableJmx", "true");
        ...
```

Later:
```
org.apache.logging.log4j.core.config.Configurator.reconfigure(builder.build());
```

## Reproduction

It's a linux jpackaged executable. Those JMX bean class-not-found-exceptions are only triggered on linux, the same code on Windows and mac doesn't even seem to miss any JMX code. The log4j2-JMX disable switches seem to have no effect. I wonder if they are even filled into the log4j component properties ?!

--- a/log4j-core/pom.xml
+++ b/log4j-core/pom.xml
@@ -70,8 +70,12 @@
       javax.lang.model.*;resolution:=optional,
       javax.tools;resolution:=optional,
       <!-- `java.sql`, which depends on `java.logging` is optional -->
+      java.sql;resolution:=optional,
       javax.sql;resolution:=optional,
       java.util.logging;resolution:=optional,
+      <!-- `java.management` is optional -->
+      java.lang.management;resolution:=optional,
+      javax.management.*;resolution:=optional,
       <!-- `java.naming` is optional -->
       javax.naming;resolution:=optional
     </bnd-extra-package-options>
@@ -88,6 +92,7 @@
       com.fasterxml.jackson.databind;transitive=false,
       com.fasterxml.jackson.dataformat.xml;transitive=false,
       com.fasterxml.jackson.dataformat.yaml;transitive=false,
+      java.management;transitive=false;static=true,
       java.naming;transitive=false,
       org.apache.commons.csv;transitive=false,
       org.fusesource.jansi;transitive=false,

--- a/log4j-core/src/main/java/org/apache/logging/log4j/core/LoggerContext.java
+++ b/log4j-core/src/main/java/org/apache/logging/log4j/core/LoggerContext.java
@@ -16,6 +16,7 @@
  */
 package org.apache.logging.log4j.core;
 
+import static org.apache.logging.log4j.core.jmx.internal.JmxUtil.isJmxDisabled;
 import static org.apache.logging.log4j.core.util.ShutdownCallbackRegistry.SHUTDOWN_HOOK_MARKER;
 
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
@@ -370,12 +371,7 @@ public class LoggerContext extends AbstractLifeCycle
             }
 
             this.setStopping();
-            try {
-                Server.unregisterLoggerContext(getName()); // LOG4J2-406, LOG4J2-500
-            } catch (final LinkageError | Exception e) {
-                // LOG4J2-1506 Hello Android, GAE
-                LOGGER.error("Unable to unregister MBeans", e);
-            }
+            unregisterJmxBeans();
             if (shutdownCallback != null) {
                 shutdownCallback.cancel();
                 shutdownCallback = null;
@@ -405,6 +401,17 @@ public class LoggerContext extends AbstractLifeCycle
         }
         LOGGER.debug("Stopped LoggerContext[name={}, {}] with status {}", getName(), this, true);
         return true;
+    }
+
+    private void unregisterJmxBeans() {
+        if (!isJmxDisabled()) {
+            try {
+                Server.unregisterLoggerContext(getName()); // LOG4J2-406, LOG4J2-500
+            } catch (final LinkageError | Exception error) {
+                // LOG4J2-1506 Hello Android, GAE
+                LOGGER.error("Unable to unregister MBeans", error);
+            }
+        }
     }
 
     /**
@@ -638,18 +645,25 @@ public class LoggerContext extends AbstractLifeCycle
 
             firePropertyChangeEvent(new PropertyChangeEvent(this, PROPERTY_CONFIG, prev, config));
 
-            try {
-                Server.reregisterMBeansAfterReconfigure();
-            } catch (final LinkageError | Exception e) {
-                // LOG4J2-716: Android has no java.lang.management
-                LOGGER.error("Could not reconfigure JMX", e);
-            }
+            registerJmxBeans();
+
             // AsyncLoggers update their nanoClock when the configuration changes
             Log4jLogEvent.setNanoClock(configuration.getNanoClock());
 
             return prev;
         } finally {
             configLock.unlock();
+        }
+    }
+
+    private static void registerJmxBeans() {
+        if (!isJmxDisabled()) {
+            try {
+                Server.reregisterMBeansAfterReconfigure();
+            } catch (final LinkageError | Exception error) {
+                // LOG4J2-716: Android has no java.lang.management
+                LOGGER.error("Could not reconfigure JMX", error);
+            }
         }
     }
 

--- a/log4j-core/src/main/java/org/apache/logging/log4j/core/jmx/Server.java
+++ b/log4j-core/src/main/java/org/apache/logging/log4j/core/jmx/Server.java
@@ -16,6 +16,8 @@
  */
 package org.apache.logging.log4j.core.jmx;
 
+import static org.apache.logging.log4j.core.jmx.internal.JmxUtil.isJmxDisabled;
+
 import java.lang.management.ManagementFactory;
 import java.util.List;
 import java.util.Map;
@@ -58,7 +60,6 @@ public final class Server {
      */
     public static final String DOMAIN = "org.apache.logging.log4j2";
 
-    private static final String PROPERTY_DISABLE_JMX = "log4j2.disable.jmx";
     private static final String PROPERTY_ASYNC_NOTIF = "log4j2.jmx.notify.async";
     private static final String THREAD_NAME_PREFIX = "jmx.notif";
     private static final StatusLogger LOGGER = StatusLogger.getLogger();
@@ -124,10 +125,6 @@ public final class Server {
             sb.append('\"');
         }
         return sb.toString();
-    }
-
-    private static boolean isJmxDisabled() {
-        return PropertiesUtil.getProperties().getBooleanProperty(PROPERTY_DISABLE_JMX, true);
     }
 
     public static void reregisterMBeansAfterReconfigure() {

--- a/log4j-core/src/main/java/org/apache/logging/log4j/core/jmx/internal/JmxUtil.java
+++ b/log4j-core/src/main/java/org/apache/logging/log4j/core/jmx/internal/JmxUtil.java
@@ -1,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.logging.log4j.core.jmx.internal;
+
+import org.apache.logging.log4j.util.PropertiesUtil;
+
+// WARNING!
+// This class must be free of any dependencies to the `java.management` module!
+// Otherwise, `isJmxDisabled()` call sites would unnecessarily require the `java.management` module.
+// For details, see: https://github.com/apache/logging-log4j2/issues/2774
+
+public final class JmxUtil {
+
+    public static boolean isJmxDisabled() {
+        return PropertiesUtil.getProperties().getBooleanProperty("log4j2.disable.jmx", true);
+    }
+
+    private JmxUtil() {}
+}

--- a/src/changelog/.2.x.x/fix_jmx_module_dep.xml
+++ b/src/changelog/.2.x.x/fix_jmx_module_dep.xml
@@ -3,6 +3,6 @@
        xmlns="https://logging.apache.org/xml/ns"
        xsi:schemaLocation="https://logging.apache.org/xml/ns https://logging.apache.org/xml/ns/log4j-changelog-0.xsd"
        type="fixed">
-  <issue id="2774" link="https://github.com/apache/logging-log4j2/issues/2774"/>
+  <issue id="2775" link="https://github.com/apache/logging-log4j2/pull/2775"/>
   <description format="asciidoc">Fix requirement on the `java.management` module when JMX is disabled, which is the default</description>
 </entry>

--- a/src/changelog/.2.x.x/fix_jmx_module_dep.xml
+++ b/src/changelog/.2.x.x/fix_jmx_module_dep.xml
@@ -1,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<entry xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xmlns="https://logging.apache.org/xml/ns"
+       xsi:schemaLocation="https://logging.apache.org/xml/ns https://logging.apache.org/xml/ns/log4j-changelog-0.xsd"
+       type="fixed">
+  <issue id="2774" link="https://github.com/apache/logging-log4j2/issues/2774"/>
+  <description format="asciidoc">Fix requirement on the `java.management` module when JMX is disabled, which is the default</description>
+</entry>

